azkaban-common/src/main/java/azkaban/executor/container/KubernetesContainerizedImpl.java

```diff
@@ -398,10 +403,71 @@ public void createContainer(final int executionId) throws ExecutorManagerExcepti
    */
   @Override
   public void deleteContainer(final int executionId) throws ExecutorManagerException {
-    deletePod(executionId);
-    if (isServiceRequired()) {
-      deleteService(executionId);
+    try { // if pod deletion is not successful, the service deletion can still be handled
+      deletePod(executionId);
+    } finally {
+      if (isServiceRequired()) {
+        deleteService(executionId);
+      }
+    }
+  }
+
+  /**
+   * This method is used to delete invalid containers in batch
+   * @param containerValidity pod validity in milliseconds
+   * @throws ExecutorManagerException
+   */
+  @Override
+  public void deleteAgedContainers(Duration containerValidity) throws ExecutorManagerException {
+    logger.info(String.format("Cleaning up containers older than %d days",
+        containerValidity.toDays()));
+    final Set<Integer> containersToDelete = getStaleContainers(containerValidity);
+    for (int execId : containersToDelete) {
+      if (execId > 0) {
+        logger.info(String.format("Deleting stale pod %s and service %s", getPodName(execId),
+            getServiceName(execId)));
+        deleteContainer(execId);
+      }
+    }
+  }
+
+  /**
+   * This method is used to fetch all stale pods in the current az cluster and namespace
+   * @param containerValidity container validity in milliseconds
+   * @return Set of stale container's execution ids
+   */
+  private Set<Integer> getStaleContainers(final Duration containerValidity) throws ExecutorManagerException {
+    Set<Integer> staleContainerExecIdSet = new HashSet<>();
+    try {
+      final String label = CLUSTER_LABEL_NAME + EQUALS_TO + this.clusterName;
+      V1PodList items= this.coreV1Api.listNamespacedPod(this.namespace, null,
+          null, null, null, label,
+          null, null, null, null);
+
+      // Get all execution ids of the pods whose age is older than Azkaban max flow running time
+      // (e.g. 10 days)
+      final Long validStartTimeStamp = System.currentTimeMillis() - containerValidity.toMillis();
+      staleContainerExecIdSet =
+          items.getItems().stream()
+              .filter((pod) -> pod.getMetadata().getCreationTimestamp().getMillis() < validStartTimeStamp)
+              .map((pod) -> pod.getMetadata().getLabels().getOrDefault(EXECUTION_ID_LABEL_NAME,
+                  EXECUTION_ID_LABEL_PREFIX + DEFAULT_EXECUTION_ID))
+              .map((str) -> str.substring(str.indexOf(
+                  EXECUTION_ID_LABEL_PREFIX) + EXECUTION_ID_LABEL_PREFIX.length()))
```

Comment:
Sorry if this comment was too late! I was wondering if we can add this label also as part of the label selector we are supplying at line 442. I believe we can pass multiple labels. I find this piece of code error-prone and slightly hard to read too.